service/src/java/org/apache/hive/service/server/HiveServer2.java

```diff
@@ -1201,63 +1204,63 @@ private static void startHiveServer2() throws Throwable {
     }
   }
 
-  private void maybeStartCompactorThreads(HiveConf hiveConf) throws Exception {
+  public Map<String, Integer> maybeStartCompactorThreads(HiveConf hiveConf) {
+    Map<String, Integer> startedWorkers = new HashMap<>();
     if (MetastoreConf.getVar(hiveConf, MetastoreConf.ConfVars.HIVE_METASTORE_RUNWORKER_IN).equals("hs2")) {
-      int numWorkers = MetastoreConf.getIntVar(hiveConf, MetastoreConf.ConfVars.COMPACTOR_WORKER_THREADS);
-      List<Map.Entry<String, String>> entries = hiveConf.getMatchingEntries(Constants.COMPACTION_POOLS_PATTERN);
+      AtomicInteger numWorkers = new AtomicInteger(MetastoreConf.getIntVar(hiveConf, MetastoreConf.ConfVars.COMPACTOR_WORKER_THREADS));
+      Map<String, Integer> customPools = CompactorUtil.getPoolConf(hiveConf);
 
       StringBuilder sb = new StringBuilder(2048);
       sb.append("This HS2 instance will act as Compactor with the following worker pool configuration:\n");
-      sb.append("Global pool size: ").append(numWorkers).append("\n");
-
-      LOG.info("Initializing the compaction pools with using the global worker limit: {} ", numWorkers);
-      while (numWorkers > 0 && entries.size() > 0) {
-        Map.Entry<String, String> entry = entries.remove(0);
-        String poolName = entry.getValue();
-        int poolWorkers = hiveConf.getInt(entry.getKey(), 0);
+      sb.append("Global pool size: ").append(numWorkers.get()).append("\n");
 
+      LOG.info("Initializing the compaction pools with using the global worker limit: {} ", numWorkers.get());
+      customPools.forEach((poolName, poolWorkers) -> {
         if (poolWorkers == 0) {
           LOG.warn("Compaction pool ({}) configured with zero workers. Skipping pool initialization", poolName);
           sb.append("Pool not initialized, 0 size: ").append(poolName).append("\n");
-          continue;
         }
-        if (poolWorkers > numWorkers) {
-          LOG.warn("Global worker pool exhausted, compaction pool ({}) will be configured with less workers than the " +
-              "required number. ({} -> {})", poolName, poolWorkers, numWorkers);
-          poolWorkers = numWorkers;
+        else if (numWorkers.get() == 0) {
```

Comment:
that should be an IllegalArgumentException

Reply:
I think this is not related to preventing compaction for selected db/tables.
If we do IllegalArgumentException here, it will be different from the intended design as described here:
https://cwiki.apache.org/confluence/display/Hive/Compaction+pooling

"_Worker allocation
The already existing hive.compactor.worker.threads configuration value holds the maximum number of worker threads. The worker allocation happens as follows:

Labeled pools are initialized in a sequential manner with random ordering.
Each pool decreases the number of available workers by its own worker count.
If the number of assignable workers is less than the configured one, the pool size will be adjusted (In other words: if the requested pool size is 5 but there are only 3 workers remaining, than the pool size will be decreased to 3).
**If the number of assignable workers is 0, the pool won’t be initialized.**
All remaining workers not used up by the labeled pools, are assigned to the default pool._"